datafusion/core/src/physical_plan/aggregates/hash.rs

```diff
@@ -257,12 +313,32 @@ fn group_aggregate_batch(
                         accumulator_set,
                         indices: vec![row as u32], // 1.3
                     };
+                    // NOTE: do NOT include the `GroupState` struct size in here because this is captured by
+                    // `group_states` (see allocation down below)
+                    allocated += group_state
```

Comment:
Figuring out how to encapsulate some of this accounting (so it wasn't inlined into the code) would make it easier to maintain I think. But I don't think that is required

Reply:
I think as long as the [`Allocator` API](https://doc.rust-lang.org/std/alloc/trait.Allocator.html) in Rust isn't stable, this will be a bit of a mess. Once it is stable, we could have very elegant memory accounting.